File: apps/api/internal/models/tool.go
package models

import (
	"encoding/json"
	"time"
)

type Tool struct {
	ID          string          `json:"id" db:"id"`
	UserID      string          `json:"userId" db:"user_id"`
	Name        string          `json:"name" db:"name"`
	DisplayName string          `json:"displayName" db:"display_name"`
	Description string          `json:"description" db:"description"`
	InputSchema json.RawMessage `json:"inputSchema" db:"input_schema"`
	Enabled     bool            `json:"enabled" db:"enabled"`
	CreatedAt   time.Time       `json:"createdAt" db:"created_at"`
	UpdatedAt   time.Time       `json:"updatedAt" db:"updated_at"`
}

type ToolBackend struct {
	ID                 string    `json:"id" db:"id"`
	ToolID             string    `json:"toolId" db:"tool_id"`
	Name               string    `json:"name" db:"name"`
	BackendType        string    `json:"backendType" db:"backend_type"`
	EndpointURL        string    `json:"endpointUrl" db:"endpoint_url"`
	AuthTokenEncrypted *string   `json:"-" db:"auth_token_encrypted"`
	AuthHeaderName     string    `json:"authHeaderName" db:"auth_header_name"`
	AuthHeaderPrefix   string    `json:"authHeaderPrefix" db:"auth_header_prefix"`
	TimeoutMs          int       `json:"timeoutMs" db:"timeout_ms"`
	Priority           int       `json:"priority" db:"priority"`
	Enabled            bool      `json:"enabled" db:"enabled"`
	CreatedAt          time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt          time.Time `json:"updatedAt" db:"updated_at"`
}

type ToolWithBackends struct {
	Tool     Tool          `json:"tool"`
	Backends []ToolBackend `json:"backends"`
}

File: apps/app/lib/api/types/tool.ts
export interface ApiTool {
  id: string;
  name: string;
  displayName: string;
  description: string;
  category: string;
  inputSchema: Record<string, unknown>;
  enabled: boolean;
  timeoutMs: number;
  maxRetries: number;
  cacheTtlSeconds: number;
  createdAt: string;
  updatedAt: string;
}

export interface ApiToolBackend {
  id: string;
  toolId: string;
  name: string;
  backendType: 'http' | 'lambda' | 'sql' | 'grpc';
  endpointUrl: string;
  httpMethod?: string;
  headers?: Record<string, string>;
  authType: 'none' | 'bearer' | 'api_key' | 'basic';
  priority: number;
  weight: number;
  timeoutMs: number;
  enabled: boolean;
  healthStatus: string;
  circuitBreakerTripped: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface ApiToolWithBackends extends Partial<ApiTool> {
  id?: string;
  name?: string;
  tool?: ApiTool;
  backends?: ApiToolBackend[];
}

export interface ApiCreateToolBackend {
  name: string;
  endpointUrl: string;
  backendType?: 'http' | 'lambda' | 'sql' | 'grpc';
  httpMethod?: string;
  headers?: Record<string, string>;
  authType?: 'none' | 'bearer' | 'api_key' | 'basic';
  authToken?: string;
  authConfig?: Record<string, string>;
  priority?: number;
  weight?: number;
  timeoutMs?: number;
}

export interface ApiCreateToolRequest {
  name: string;
  displayName?: string;
  description?: string;
  category?: string;
  inputSchema?: Record<string, unknown>;
  enabled?: boolean;
  timeoutMs?: number;
  maxRetries?: number;
  cacheTtlSeconds?: number;
  backends?: ApiCreateToolBackend[];
}

export interface ApiToolExecutionResult {
  toolId: string;
  backendId?: string;
  success: boolean;
  latencyMs: number;
  statusCode?: number;
  result?: unknown;
  error?: string;
  cached: boolean;
}
